pkg/mgmt/snapshot/snapshot.go

```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2021 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package snapshot
+
+import (
+	"fmt"
+	"time"
+
+	apis "github.com/openebs/lvm-localpv/pkg/apis/openebs.io/lvm/v1alpha1"
+	lvm "github.com/openebs/lvm-localpv/pkg/lvm"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// isDeletionCandidate checks if a lvm snapshot is a deletion candidate.
+func (c *SnapController) isDeletionCandidate(snap *apis.LVMSnapshot) bool {
+	return snap.ObjectMeta.DeletionTimestamp != nil
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two.
+func (c *SnapController) syncHandler(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the snap resource with this namespace/name
+	snap, err := c.snapLister.LVMSnapshots(namespace).Get(name)
+	if k8serror.IsNotFound(err) {
+		runtime.HandleError(fmt.Errorf("lvm snapshot '%s' has been deleted", key))
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	snapCopy := snap.DeepCopy()
+	err = c.syncSnap(snapCopy)
+	return err
+}
+
+// enqueueSnap takes a LVMSnapshot resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than LVMSnapshot.
+func (c *SnapController) enqueueSnap(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// synSnap is the function which tries to converge to a desired state for the
+// LVMSnapshot
+func (c *SnapController) syncSnap(snap *apis.LVMSnapshot) error {
+	var err error
+	// LVMSnapshot should be deleted. Check if deletion timestamp is set
+	if c.isDeletionCandidate(snap) {
+		err = lvm.DestroySnapshot(snap)
+		if err == nil {
+			lvm.RemoveSnapFinalizer(snap)
+		}
+	} else {
+		// if finalizer is not set then it means we are creating
+		// the lvm snapshot.
+		if snap.Finalizers == nil {
```

Comment:
can we rely on snap.Spec.Status to be Pending here?

Reply:
@pawanpraka1  Which is the case that we are hitting here?

Reply:
rather than checking whether finalizer us set it is better to check if request is pending then go ahead and create the snapshot.

Reply:
Yep. we can do that. Why is this approach not followed in zfs localPV?

Reply:
Got the answer : "as there was no status initially, finalizer was used to check if the snapshot was created."

Updated the changes.